### PR TITLE
[SMR-2019] Upgrade the deployment scripts to use a hardware ledger

### PR DESCRIPTION
**Pull request by shirren**
# Summary

[SMR-2019] Upgrade the deployment scripts to use a hardware ledger.

[SMR-2019]: https://immutable.atlassian.net/browse/SMR-2019?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by wcgcyx — COMMENTED** [review]
I'm wondering if we have done a test run on testnet/devnet yet? One potential concern from me is that we probably need to override the gas options when transacting on zkEVM.

**Comment by shirren** [comment]
> I'm wondering if we have done a test run on testnet/devnet yet? One potential concern from me is that we probably need to override the gas options when transacting on zkEVM.

@wcgcyx we plan to do a test run on devnet and testnet before running on mainnet. I think it might work because these scripts were run on devnet, and testnet before 🤞🏽 . The only significant change is the introduction of the ledger.

File: hardhat.config.ts [hunk 1]
```diff
@@ -33,66 +30,36 @@ const config: HardhatUserConfig = {
...
-      url: 'http://localhost:8555'
-    }
-  },
-  etherscan: {
-    // Your API key for Etherscan
-    // Obtain one at https://etherscan.io/
-    apiKey: networkConfig('mainnet').etherscan
+    localhost: {
+      url: 'http://127.0.0.1:8545',
+      accounts: [process.env.DEPLOYER_PRIV_KEY!, process.env.WALLET_IMPL_CHANGER_PRIV_KEY!]
```
Comment (igx-C) [inline hardhat.config.ts:39]: Do we plan to use an existing privileged transaction multisig for the `WALLET_IMPL_CHANGER_PRIV_KEY`

File: scripts/archived/index.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
+import * as fs from 'fs';
```
Comment (igx-C) [inline scripts/archived/index.ts:1]: nit; is there a reason to keep the scripts in `archived`? I don't see us using them in our main scripts.

File: scripts/README.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+5. Set the deployer key to be a unique value for the run.
+6. Execute the command `npx npm run deployToLocal`
+    1. IMPORTANT: Remember to copy the correct `.env.X` file into `.env` before execution.
+7. Edit the following environment variables in the Relayer in the `.env.local` and `local-deployment.yaml` files
+    1. DEPLOY_AND_EXECUTE_ADDRESS to equal the address of the `MultiCallDeploy`
+    2. FACTORY_ADDRESS to equal the address of the `Factory`
+    3. MAIN_MODULE_ADDRESS to equal the address of the `StartupWalletImpl`
+    4. IMMUTABLE_SIGNER_CONTRACT_ADDRESS to equal the address of the `ImmutableSigner`
+
+# Steps for deploying to Devnet
```
Comment (igx-C) [inline scripts/README.md:18]: Would these be the same steps we'd be running on Testnet and Mainnet but just changing `dev`?

File: scripts/README.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+1. If the `OwnableCreate2Deployer` contract has not been deployed, deploy it using the ledger. The Ledger is now the owner of the contract, and the
+only entity allowed to call the deploy function on the contract;
+    1. forge create --rpc-url <https://rpc.dev.immutable.com> --constructor-args "<ADDRESS_OF_LEDGER_ACCOUNT>" --legacy --hd-path "m/44'/60'/0'/0/0" src/OwnableCreate2Deployer.sol:OwnableCreate2Deployer.
+2. Set the value of DEPLOYER_CONTRACT_ADDRESS in the environment to equal the address of the OwnableCreate2Deployer contract deployed of step 1 above.
+3. Set the deployer key to be a unique value for the run.
+4. Execute the command `npx npm run deployToDevnet`.
+    1. IMPORTANT: Remember to copy the correct `.env.X` file into `.env` before execution.
+5. Edit the following environment variables in the Relayer in the `.env.local` and `local-deployment.yaml` files
```
Comment (ermyas) [inline scripts/README.md:27]: should these be `.env.devnet` and `devnet-deployment.yaml`?

File: scripts/environment.ts [hunk 5]
```diff
@@ -0,0 +1,31 @@
+import * as path from 'path';
+
+/**
+ * An environment should include the address details for the submitter,
+ * and signer address.
+ */
+export interface EnvironmentInfo {
+  submitterAddress?: string;
+  signerAddress?: string;
```
Comment (igx-C) [inline scripts/environment.ts:9]: Is it a problem if neither are provided?

File: scripts/wallet-options.ts [hunk 6]
```diff
@@ -0,0 +1,54 @@
+import { ethers as hardhat } from 'hardhat';
+import { Signer } from 'ethers';
+import { LedgerSigner } from './ledger-signer';
+import { EnvironmentInfo } from './environment';
+
+const mainnetEnv = 'mainnet';
```
Comment (igx-C) [inline scripts/wallet-options.ts:6]: Unused?

File: scripts/deploy.ts [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+  console.log(
+    `[${network}] Wallet Impl Locator Changer Address: ${await wallets.getWalletImplLocatorChanger().getAddress()}`
+  );
+
+  console.log(`[${network}] Deploying contracts...`);
+
+  // Key for the salt, use this to change the address of the contract
+  let key: string = 'relayer-deployer-key-1';
+
+  // 1. Deploy multi call deploy
```
Comment (igx-C) [inline scripts/deploy.ts:88]: For posterity, could we make a note of which contract must be deployed with the nonce reserver and the rest by the contract factory?

File: scripts/wallet-options.ts [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+  private useLedger: boolean;
+  private ledger: LedgerSigner;
+  private coldWallet: Signer;
+  private walletImplLocatorImplChanger: Signer;
+
+  constructor(env: EnvironmentInfo, coldWallet: Signer, walletImplLocatorImplChanger: Signer) {
+    console.log(`[${env.network}] Using ledger for operations...`);
+    this.useLedger = true;
+    const accountIndex0 = 0;
+    const derivationPath0 = `m/44'/60'/${accountIndex0.toString()}'/0/0`;
```
Comment (ermyas) [inline scripts/wallet-options.ts:23]: can i confirm that assuming account index 0 here in all instances is fine?

File: scripts/deploy.ts [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+  const multiCallDeploy = await deployContract(env, wallets, key, 'MultiCallDeploy', [multiCallAdminPubKey, submitterAddress]);
+
+  // 2. Deploy factory with multi call deploy address as deployer role EST
```
Comment (igx-C) [inline scripts/deploy.ts:91]: @shirren, am I reading this correctly that every contract using `deployContract` will use the ledger-based deployer key and our OwnableCreate2Deployer to deploy said contract? Can you point me to where we're using the nonce reserver / switch to a regular deployment approach?

**Review by ermyas — APPROVED** [review]

File: hardhat.config.ts (hunk 1 above)
Comment (shirren) [reply hardhat.config.ts:39]: Yes.

File: scripts/README.md (hunk 3 above)
Comment (shirren) [reply scripts/README.md:18]: Yup. I've not included them just yet. I just wanted to run the dev deployment first, then update it with instructions incase I missed something here.

File: scripts/archived/index.ts (hunk 2 above)
Comment (shirren) [reply scripts/archived/index.ts:1]: After a successful deployment, I am going to cleanup the archive folder once and for all 🙏🏽

File: scripts/deploy.ts (hunk 9 above)
Comment (shirren) [reply scripts/deploy.ts:91]: I thought we were going to use factory for all of them, why do we need the nonce receiver?

File: scripts/environment.ts (hunk 5 above)
Comment (shirren) [reply scripts/environment.ts:9]: It is a problem, but `process.env.` returns `string | undefined` hence the type. So we could do the same thing as we do below, which does make more sense.

File: scripts/wallet-options.ts (hunk 8 above)
Comment (shirren) [reply scripts/wallet-options.ts:23]: Nope. The script is very simple. I will set this to the index of the corresponding ledger account once, and for all for deployment.

**Pull request merged**